eth/handler.go

```diff
@@ -608,7 +608,13 @@ func (pm *ProtocolManager) handleMsg(p *peer) error {
 				return errResp(ErrDecode, "msg %v: %v", msg, err)
 			}
 			// Retrieve the requested state entry, stopping if enough was found
-			if entry, err := pm.blockchain.TrieNode(hash); err == nil {
+			// todo now the code and trienode is mixed in the protocol level,
+			// separate these two types.
+			entry, err := pm.blockchain.TrieNode(hash)
+			if len(entry) == 0 || err != nil {
+				entry, err = pm.blockchain.ContractCode(hash)
+			}
```

Comment:
This is a bit awkward. The `TrieNode(hash)` will: 
1. Check `cleans[hash]`, 
2. check `dirty[hash]` 
3. and finally `rawdb.ReadTrieNode(hash)`, which does `db.Get(hash.Bytes())`

Then we come to `ContractCode(hash), which will : 
1.  Check `cleans[hash]`, 
2.  check `codes[hash]` , 
3.  and finally ` rawdb.ReadCode(db.diskdb, hash)`, which will do
  a) `db.Get(hash)` followed by 
  b)` db.Get(prefix .. hash)`. 


This means that we'll do `3` separate disk-lookups for code that is stored in the new format. Those will be cached in `cleans`, but if someone is requesting a state entry we don't have, we'll do three disk lookups before we can say "sorry, don't have it". 

There's got to be some better way of doing this, perhaps by having something which semantically means `pm.blockchain.ContractCode-please-But-Dont-Bother-Look-it-up-Up-As-A-TrieNode_kthxbye`

Reply:
To follow up on this, can e.g. `pm.blockchain.ContractCode` take a boolean to indicate whether it should also consult the non-prefixed data?

Reply:
In this case, we need a background upgrader to convert the original scheme. 
If it's done, then we don't need to consult the non-prefixed at all.

So here is a tradeOff. 
- Should we implement the upgrade path?
- In this API, the possibility of retrieving code with 3 db reads is quite low
   - I guess 90% APIs are used to retrieve trie nodes
   - I guess 99% APIs will only need 2 reads for retrieving codes(with no prefix)
- But in order to reach our final goal(prefix the code, we need the upgrader anyway


How should we implement the upgrader?
- User needs to activate the snapshot(for state iteration)
- Iterate the snapshot, convert all non-prefixed code with new scheme

Can we do it now?
- Snapshot is not enabled by default
- Snapshot relevant code is not 100% production ready

Reply:
> * This PR separates the code, but in practice, it's noop, no legacy data will be converted(missing upgrader)

I disagree. From time to time, node operators do a full resync, if nothing else than because it cleans out some data. That means those users _will_ be upgraded 100% to the new scheme